packages/expo-updates/android/src/main/java/expo/modules/updates/loader/Crypto.kt

```diff
@@ -104,4 +123,105 @@ object Crypto {
+      }
+
+      certificate
+    }
+
+    val algorithm: CodeSigningAlgorithm by lazy {
+      CodeSigningAlgorithm.parseFromString(metadata?.get(CODE_SIGNING_METADATA_ALGORITHM_KEY))
+    }
+
+    val keyId: String by lazy {
+      metadata?.get(CODE_SIGNING_METADATA_KEY_ID_KEY) ?: CODE_SIGNING_METADATA_DEFAULT_KEY_ID
+    }
+  }
+
+  fun createAcceptSignatureHeader(codeSigningConfiguration: CodeSigningConfiguration): String {
+    return Dictionary.valueOf(
+      mapOf(
+        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_SIGNATURE to BooleanItem.valueOf(true),
+        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_KEY_ID to StringItem.valueOf(codeSigningConfiguration.keyId),
+        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_ALGORITHM to StringItem.valueOf(codeSigningConfiguration.algorithm.algorithmName)
+      )
+    ).serialize()
+  }
+
+  data class SignatureHeaderInfo(val signature: String, val keyId: String, val algorithm: CodeSigningAlgorithm)
+
+  fun parseSignatureHeader(signatureHeader: String?): SignatureHeaderInfo {
+    if (signatureHeader == null) {
+      throw Error("No expo-signature header specified")
+    }
+
+    val signatureMap = Parser(signatureHeader).parseDictionary().get()
+
+    val sigFieldValue = signatureMap[CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_SIGNATURE]
+    val keyIdFieldValue = signatureMap[CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_KEY_ID]
+    val algFieldValue = signatureMap[CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_ALGORITHM]
+
+    val signature = if (sigFieldValue is StringItem) {
+      sigFieldValue.get()
+    } else throw Error("Structured field $CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_SIGNATURE not found in expo-signature header")
+    val keyId = if (keyIdFieldValue is StringItem) {
+      keyIdFieldValue.get()
+    } else CODE_SIGNING_METADATA_DEFAULT_KEY_ID
+    val alg = if (algFieldValue is StringItem) {
+      algFieldValue.get()
+    } else null
+
+    return SignatureHeaderInfo(signature, keyId, CodeSigningAlgorithm.parseFromString(alg))
+  }
+
+  fun isSignatureValid(configuration: CodeSigningConfiguration, info: SignatureHeaderInfo, bytes: ByteArray): Boolean {
+    // check that the key used to sign the response is the same as the key embedded in the configuration
+    // TODO(wschurman): this may change for child certificates and development certificates
+    if (info.keyId != configuration.keyId) {
+      throw Error("Key with keyid=${info.keyId} from signature not found in client configuration")
+    }
+
+    // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
+    // algorithm specified in the configuration
+    if (info.algorithm != configuration.algorithm) {
+      Log.i(TAG, "Key with alg=${info.algorithm} from signature does not match client configuration algorithm, continuing")
+    }
+
+    return Signature.getInstance(
+      when (configuration.algorithm) {
```

Comment:
It's a bit tricky when there's only one algorithm but think it's a good idea to set things up so it becomes possible to write a test case ensuring that configuration.algorithm is used and info.algorithm is ignored.

Reply:
It's somewhat set up this way, but is pretty tough to test at this point since the default is the same as the only supported algorithm.